File: day-3a/src/main.py
import re
from collections.abc import Iterator


def gather_surrounding_chars(
    loc_range: range,
    *,
    lineno: int,
    line: str,
    all_lines: list[str]
) -> Iterator[str]:
    left = max(0, loc_range.start - 1)
    right = min(len(all_lines[0]) - 1, loc_range.stop)

    try:
        prev_line = all_lines[lineno - 1]
    except IndexError:
        pass
    else:
        yield from prev_line[left:(right+1)]

    try:
        next_line = all_lines[lineno + 1]
    except IndexError:
        pass
    else:
        yield from next_line[left:(right+1)]

    if not line[left].isdigit():
        yield line[left]
    if not line[right].isdigit():
        yield line[right]


def char_is_symbol(char: str) -> bool:
    return char != "." and not char.isdigit()


def is_part_number(
    loc_range: range, lineno: int, line: str, all_lines: list[str]
) -> bool:
    surrounding_chars = gather_surrounding_chars(
        loc_range, lineno=lineno, line=line, all_lines=all_lines
    )
    return any(char_is_symbol(char)for char in surrounding_chars)


def gather_part_numbers_from_line(
    lineno: int, line: str, all_lines: list[str]
) -> Iterator[int]:
    for match in re.finditer(r"\d+", line):
        loc_range = range(match.start(), match.end())
        if is_part_number(loc_range, lineno, line, all_lines):
            yield int(match[0])


def gather_part_numbers_from_file(lines: list[str]) -> Iterator[int]:
    for i, line in enumerate(lines):
        yield from gather_part_numbers_from_line(i, line, lines)


def solve(filename: str) -> int:
    with open(filename) as f:
        lines = f.read().splitlines()
    return sum(gather_part_numbers_from_file(lines))


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()

File: day-3a/src/main.rs
use std::cmp::{max, min};
use std::fs::read_to_string;
use std::iter::Extend;
use std::ops::Not;
use regex::Regex;

struct LocRange {
    start: usize,
    end: usize
}

fn gather_surrounding_chars(
    loc_range: LocRange,
    lineno: usize,
    line: &str,
    all_lines: &Vec<&str>
) -> Vec<char> {
    let mut answer = Vec::new();
    let left = max(0, loc_range.start.wrapping_sub(1));
    let right = min(all_lines[0].len(), loc_range.end);
    match all_lines.get(lineno.wrapping_sub(1)) {
        Some(prev_line) => answer.extend(prev_line[left..right].chars()),
        None => {}
    }
    match all_lines.get(lineno.wrapping_add(1)) {
        Some(next_line) => answer.extend(next_line[left..right].chars()),
        None => {}
    }
    let line_as_bytes = line.as_bytes();
    for c in [line_as_bytes[left], line_as_bytes[right.wrapping_sub(1)]] {
        if c.is_ascii_digit().not() {
            answer.push(c as char)
        }
    };
    answer
}

fn char_is_symbol(c: &char) -> bool{
    let period: &char = &'.';
    c.is_digit(10) && c != period
}

fn is_part_number(loc_range: LocRange, lineno: usize, line: &str, all_lines: &Vec<&str>) -> bool {
    let surrounding_chars = gather_surrounding_chars(loc_range, lineno, line, all_lines);
    surrounding_chars.iter().any(char_is_symbol)
}

fn gather_part_numbers_from_line(lineno: usize, line: &str, all_lines: &Vec<&str>) -> Vec<u32> {
    let mut answer = Vec::new();
    let number_re = Regex::new(r"\d+").unwrap();
    for number_match in number_re.find_iter(line) {
        let loc_range = LocRange{start: number_match.start(), end: number_match.end()};
        if is_part_number(loc_range, lineno, line, all_lines) {
            let parsed_number = number_match.as_str().parse::<u32>().unwrap();
            answer.push(parsed_number)
        }
    }
    answer
}

fn gather_part_numbers_from_file(lines: Vec<&str>) -> Vec<u32> {
    let mut answer = Vec::new();
    for (lineno, line) in lines.iter().enumerate() {
        let found_parts = gather_part_numbers_from_line(lineno, line, &lines);
        answer.extend(found_parts);
    };
    answer
}

fn solve(filename: &str) -> u32 {
    let file = read_to_string(filename).unwrap();
    let lines: Vec<&str> = file.lines().collect();
    gather_part_numbers_from_file(lines).iter().sum()
}

fn main() {
    let answer = solve("src/input.txt");
    println!("{}", answer)
}